%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Masters/Doctoral Thesis 
% LaTeX Template
% Version 2.3 (25/3/16)
%
% This template has been downloaded from:
% http://www.LaTeXTemplates.com
%
% Version 2.x major modifications by:
% Vel (vel@latextemplates.com)
%
% This template is based on a template by:
% Steve Gunn (http://users.ecs.soton.ac.uk/srg/softwaretools/document/templates/)
% Sunil Patel (http://www.sunilpatel.co.uk/thesis-template/)
%
% Template license:
% CC BY-NC-SA 3.0 (http://creativecommons.org/licenses/by-nc-sa/3.0/)
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%----------------------------------------------------------------------------------------
%	PACKAGES AND OTHER DOCUMENT CONFIGURATIONS
%----------------------------------------------------------------------------------------

\documentclass[
11pt, % The default document font size, options: 10pt, 11pt, 12pt
%oneside, % Two side (alternating margins) for binding by default, uncomment to switch to one side
%chapterinoneline,% Have the chapter title next to the number in one single line
english, % ngerman for German
singlespacing, % Single line spacing, alternatives: onehalfspacing or doublespacing
%draft, % Uncomment to enable draft mode (no pictures, no links, overfull hboxes indicated)
%nolistspacing, % If the document is onehalfspacing or doublespacing, uncomment this to set spacing in lists to single
%liststotoc, % Uncomment to add the list of figures/tables/etc to the table of contents
%toctotoc, % Uncomment to add the main table of contents to the table of contents
%parskip, % Uncomment to add space between paragraphs
%nohyperref, % Uncomment to not load the hyperref package
headsepline, % Uncomment to get a line under the header
]{main} % The class file specifying the document structure

\usepackage[utf8]{inputenc} % Required for inputting international characters
\usepackage[T1]{fontenc} % Output font encoding for international characters
\usepackage{palatino} % Use the Palatino font by default
\usepackage[backend=bibtex,style=authoryear,natbib=true]{biblatex} % Use the bibtex backend with the authoryear citation style (which resembles APA)
\usepackage[autostyle=true]{csquotes} % Required to generate language-dependent quotes in the bibliography

%----------------------------------------------------------------------------------------
%	RapMap Packages
%----------------------------------------------------------------------------------------
\usepackage[justification=centering]{caption}
\usepackage{xspace}
\usepackage{bm}
% \usepackage{xifthen}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amstext}
\usepackage{amsthm}
% \usepackage{graphicx}
% \usepackage{booktabs}
% \usepackage[normalem]{ulem}
% \usepackage{setspace}
\usepackage{tabulary}
\usepackage{tablefootnote}
% \usepackage{float}
% \usepackage{natbib}
% \usepackage{newfloat}
\usepackage{cleveref}
\usepackage{footmisc}

%----------------------------------------------------------------------------------------
%	MARGIN SETTINGS
%----------------------------------------------------------------------------------------
\addbibresource{main.bib} % The filename of the bibliography

\geometry{
	paper=a4paper, % Change to letterpaper for US letter
	inner=2.5cm, % Inner margin
	outer=3.8cm, % Outer margin
	bindingoffset=2cm, % Binding offset
	top=1.5cm, % Top margin
	bottom=1.5cm, % Bottom margin
	%showframe,% show how the type block is set on the page
}

%----------------------------------------------------------------------------------------
%	THESIS INFORMATION
%----------------------------------------------------------------------------------------

\thesistitle{Efficient quantification of bulk and droplet based single-cell RNA-sequencing data}
% Your thesis title, this is used in the title and abstract, print it elsewhere with \ttitle
\supervisor{Prof. Rob \textsc{Patro} } % Your supervisor's name, this is used in the title page, print it elsewhere with \supname
\examinerO{Prof. Michael \textsc{Bender}} % Your examiner's name, this is not currently used anywhere in the template, print it elsewhere with \examname
\examinerTh{Prof. Steve \textsc{Skiena}} % Your examiner's name, this is not currently used anywhere in the template, print it elsewhere with \examname
\degree{Doctor of Philosophy} % Your degree name, this is used in the title page and abstract, print it elsewhere with \degreename
\author{Avi \textsc{Srivastava}} % Your name, this is used in the title page and abstract, print it elsewhere with \authorname
\addresses{} % Your address, this is not currently used anywhere in the template, print it elsewhere with \addressname

\subject{Computational Biology} % Your subject area, this is not currently used anywhere in the template, print it elsewhere with \subjectname
\keywords{RNA-seq, dscRNA-seq, quantification, quasi-mapping, UMI-deduplication} % Keywords for your thesis, this is not currently used anywhere in the template, print it elsewhere with \keywordnames
\university{\href{http://www.stonybrook.edu/}{Stony Brook University}} % Your university's name and URL, this is used in the title page and abstract, print it elsewhere with \univname
\department{\href{https://www.cs.stonybrook.edu/}{Department of Computer Science}} % Your department's name and URL, this is used in the title page and abstract, print it elsewhere with \deptname
\group{\href{http://combine-lab.github.io/}{Combine-Lab}} % Your research group's name and URL, this is used in the title page, print it elsewhere with \groupname
\faculty{\href{http://www.robpatro.com/redesign/}{Rob Patro}} % Your faculty's name and URL, this is used in the title page and abstract, print it elsewhere with \facname

\hypersetup{pdftitle=\ttitle} % Set the PDF's title to your title
\hypersetup{pdfauthor=\authorname} % Set the PDF's author to your name
\hypersetup{pdfkeywords=\keywordnames} % Set the PDF's keywords to your keywords

\begin{document}

\frontmatter % Use roman page numbering style (i, ii, iii, iv...) for the pre-content pages

\pagestyle{plain} % Default to the plain heading style until the thesis style is called for the body content

%----------------------------------------------------------------------------------------
%	TITLE PAGE
%----------------------------------------------------------------------------------------

\begin{titlepage}
\begin{center}

{\scshape\LARGE \univname\par}\vspace{1.5cm} % University name
\textsc{\Large Preliminary Examination --- Defense Proposal}\\[0.5cm] % Thesis type

\HRule \\[0.4cm] % Horizontal line
{\huge \bfseries \ttitle\par}\vspace{0.4cm} % Thesis title
\HRule \\[1.5cm] % Horizontal line
 
\begin{minipage}[t]{0.4\textwidth}
\begin{flushleft} \large
\emph{Author:}\\
\href{http://k3yavi.github.io/}{\authorname} % Author name - remove the \href bracket to remove the link
\end{flushleft}
\end{minipage}
\begin{minipage}[t]{0.4\textwidth}
\begin{flushright} \large
\emph{Committee:} \\
\href{http://www.robpatro.com/redesign/}{\supname  (Advisor)}\\ % Supervisor name - remove the \href bracket to remove the link 
\href{http://www3.cs.stonybrook.edu/~bender/}{\examinerOname}\\
\href{https://www3.cs.stonybrook.edu/~skiena/}{\examinerThname}
\end{flushright}
\end{minipage}\\[3cm]
 
\large \textit{A Preliminary report submitted in partial fulfillment of the requirements\\ for the degree of \degreename}\\[0.3cm] % University requirement text
\textit{in the}\\[0.4cm]
\groupname\\\deptname\\[2cm] % Research group name and department name
 
{\large \today}\\[4cm] % Date
%\includegraphics{Logo} % University/department logo - uncomment to place it
 
\vfill
\end{center}
\end{titlepage}

%----------------------------------------------------------------------------------------
%	DECLARATION PAGE
%----------------------------------------------------------------------------------------

% \begin{declaration}
% \addchaptertocentry{\authorshipname}
%
% \noindent I, \authorname, declare that this report titled, \enquote{\ttitle} and the work presented in it are my own. I confirm that:
%
% \begin{itemize}
% \item This work was done wholly or mainly while in candidature for a research degree at this University.
% \item Where any part of this thesis has previously been submitted for a degree or any other qualification at this University or any other institution, this has been clearly stated.
% \item Where I have consulted the published work of others, this is always clearly attributed.
% \item Where I have quoted from the work of others, the source is always given. With the exception of such quotations, this thesis is entirely my own work.
% \item I have acknowledged all main sources of help.
% \item Where the thesis is based on work done by myself jointly with others, I have made clear exactly what was done by others and what I have contributed myself.\\
% \end{itemize}
%
% \noindent Signed:\\
% \rule[0.5em]{25em}{0.5pt} % This prints a line for the signature
%
% \noindent Date:\\
% \rule[0.5em]{25em}{0.5pt} % This prints a line to write the date
% \end{declaration}
%
% \cleardoublepage

%----------------------------------------------------------------------------------------
%	QUOTATION PAGE
%----------------------------------------------------------------------------------------

\vspace*{0.2\textheight}

\noindent\enquote{\itshape Nature uses only the longest threads to weave her patterns, so that each small piece of her fabric reveals the organization of the entire tapestry.}\bigbreak

\hfill Richard Feynman

%----------------------------------------------------------------------------------------
%	ABSTRACT PAGE
%----------------------------------------------------------------------------------------

\begin{abstract}
\addchaptertocentry{\abstractname} % Add the abstract to the table of contents

RNA Sequencing (RNA-seq) technologies are evolving rapidly, and with them, the requirement 
for fast and accurate tools to analyze the generated data. The first step of many RNA-seq
analyses requires us to solve the problem of \ra. When reads are aligned to a collection of 
reference sequences that share a substantial amount of sub-sequence (near or exact repeats), 
a single read can have many potential alignments, and considering all such alignment can be 
crucial for downstream analysis like quantification. This study discusses the evolution of 
the tools developed to handle the problem of \ra for next generation sequencing, their impact 
on RNA-seq analysis, introduce a novel concept, \qm, and an efficient algorithm implementing 
this approach called \rapmap which maps RNA-seq \reads (sequences) to the reference sequence(s) 
(transcriptome). By attempting only to report the potential loci of origin of a sequencing read, 
and not the base-to-base alignment by which it derives from the reference, \rapmap is capable of 
\textit{mapping} sequencing reads to a target transcriptome substantially faster than existing 
alignment tools. The algorithm we employ to implement \qm uses several efficient data structures 
and takes advantage of the special structure of shared sequence prevalent in transcriptomes to 
rapidly provide highly-accurate mapping information.

Bulk RNA-sequencing is an established method to perform genome-wide quantification of gene 
expressions. However, quantification tools for bulk RNA-seq cannot be directly used for 
droplet based single cell RNA-sequencing (dscRNA-seq) data. We introduce alevin, a fast 
end-to-end pipeline to process droplet-based single-cell RNA sequencing data, performing 
cell barcode detection, read mapping, unique molecular identifier (UMI) deduplication, 
gene count estimation, and cell barcode whitelisting. Alevin’s approach to UMI deduplication 
considers transcript-level constraints on the molecules from which UMIs may have arisen and 
accounts for both gene-unique reads and reads that multimap between genes. This addresses 
the inherent bias in existing tools which discard gene-ambiguous reads and improves the 
accuracy of gene abundance estimates. Alevin is considerably faster, typically 8 times, 
than existing gene-quantification approaches, while also using less memory.

Finally, we propose to study the effectiveness of both our methods and extend it further.
\qm as concept is a novel but it trades-off speed with accuracy, which shows in 
\textit{relatively} complicated datasets. In our preliminary study for the extention of 
\qm, we extensively explore the idea of selectively aligning (SA) reads based on
their genotypic nature and optimize \rapmap for efficient dynamic dataset dependent speed
and accurcay tradeoff. \Alevin utilizes \rapmap for transcriptome mapping and improved
mapping would be directly reflected on the performance of \alevin. Moreover, we propose
a principled approach and study the affect of further extending \alevin's UMI 
deduplication algorithm to include unspliced RNA.

\end{abstract}

%----------------------------------------------------------------------------------------
%	ACKNOWLEDGEMENTS
%----------------------------------------------------------------------------------------

\begin{acknowledgements}
\addchaptertocentry{\acknowledgementname} % Add the acknowledgements to the table of contents

My experience as a Ph.D. student at Stony Brook has been life changing as it has helped me 
grow both personally and professionally. A big fraction of this is due to my advisor, 
Rob Patro, whom I'd like to thank for his patience, guidance and encouragement for letting 
me explore the field of computational biology has I did not had any previous experience. 
Working with him, has been very exciting; He taught me how to do science and his vision 
and tenaciousness to work and improve the RNA-seq quantification has helped me learn, 
how looking at a problem from many different angles generate exciting new ideas. I also 
feel, his love for statically-typed programmed programming languages has helped me realize 
the importance of designing efficient tools which I hope to keep for the rest of my career.

I'd like to thank other members of my committee, Steve Skiena and Michael Bender whose 
important contributions through various lectures, talks and feedback on this thesis has 
been pivotal for my cademic development while at Stony Brook university. I'm honored to 
have them as my committee members.

I'm very grateful to my colleagues and it was an exciting experience to learn and grow
alongside them over the years. In particular Laraib Malik, Hirak sarkar, Mohsen Zakeri
and Fatemeh Almodaresi made Combine-Lab an interesting place to work and provide
insightful discussions. I couldn't imagine working through all the paper-deadlines
had been more enjoyable and pleasant experience without them. I'd also like to thank Mike Love, 
Charlotte Soneson, Geet Duggal, Richard Smith-Unna, Owen Dando, Darya Filipova, Patrick Marks 
and Paul Ryvkin for their insightful discussions regarding various aspects of the work.

I couldn't thank enough to my family for their support throughout my journey of learning. 
Specially my father, Y K Srivastav, who guided from my childhood to learn and motivated me 
to think out-of-box. I'd like to thank my mother Ritima Srivastav, brother Prakhar Srivastava,
late grand parents, Govind Prasad Srivastava, Savitri Srivastava and all cousins brother and
sister for providing comfortable and loving environment they provided for me growing up. 

At last, I would also like to thank Stony Brook Research Computing and 
Cyberinfrastructure, and the Institute for Advanced Computational Science at Stony Brook 
University for access to the high-performance SeaWulf computing system, which was made 
possible by a $\$1.4$M National Science Foundation grant (\#1531492).
\end{acknowledgements}

%----------------------------------------------------------------------------------------
%	LIST OF CONTENTS/FIGURES/TABLES PAGES
%----------------------------------------------------------------------------------------

\tableofcontents % Prints the main table of contents

\listoffigures % Prints the list of figures

\listoftables % Prints the list of tables

% ----------------------------------------------------------------------------------------
% 	ABBREVIATIONS
% ----------------------------------------------------------------------------------------

\begin{abbreviations}{ll} % Include a list of abbreviations (a table of two columns)

\textbf{ARD} & \textbf{A}bsolute  \textbf{R}elative \textbf{D}ifference\\
\textbf{BLAST} & \textbf{B}asic \textbf{L}ocal \textbf{A}lignment \textbf{S}earch \textbf{T}ool\\
\textbf{BWT} & \textbf{B}urrows-\textbf{W}heeler \textbf{T}ransform\\
\textbf{DFS} & \textbf{D}epth-\textbf{F}irst \textbf{S}earch\\
\textbf{DP} & \textbf{D}ynamic-\textbf{P}rogram\\
\textbf{Indel} & \textbf{In}setion (and) \textbf{D}eletion\\
\textbf{MARD} & \textbf{M}ean \textbf{A}bsolute  \textbf{R}elative \textbf{D}ifference\\
\textbf{NGS} & \textbf{N}ext \textbf{G}eneration \textbf{S}equencing\\
\textbf{SA} & \textbf{S}uffix \textbf{A}rray\\
\textbf{SGS} & \textbf{S}econd \textbf{G}eneration \textbf{S}equencing\\
\textbf{SIMD} & \textbf{S}ingle \textbf{I}nstruction \textbf{M}ultiple \textbf{D}ata\\
\textbf{SMEM} & \textbf{S}uper \textbf{M}aximal \textbf{E}xact \textbf{M}atch\\
\textbf{T-DBG} & \textbf{T}ranscriptome - \textbf{d}e \textbf{B}ruijn \textbf{G}raph\\
\textbf{TPEF} & \textbf{T}rue \textbf{P}ositive \textbf{E}rror \textbf{F}raction\\
\textbf{TPME} & \textbf{T}rue \textbf{P}ositive  \textbf{M}edian \textbf{E}rror\\
\textbf{VBEM} & \textbf{V}ariational \textbf{B}ayesian - \textbf{E}xpectation \textbf{M}aximization\\
\textbf{wMARD} & \textbf{w}eighted \textbf{M}ean \textbf{A}bsolute  \textbf{R}elative \textbf{D}ifference\\
\textbf{dscRNA-seq} & \textbf{d}roplet based \textbf{s}ingle \textbf{c}ell  \textbf{RNA}-seq\\
\textbf{CB} & \textbf{C}ellular \textbf{B}arcode\\
\textbf{UMI} & \textbf{U}nique \textbf{M}olecule \textbf{I}dentifier\\

\end{abbreviations}

%----------------------------------------------------------------------------------------
%	PHYSICAL CONSTANTS/OTHER DEFINITIONS
%----------------------------------------------------------------------------------------

% \begin{constants}{lr@{${}={}$}l} % The list of physical constants is a three column table
%
% % The \SI{}{} command is provided by the siunitx package, see its documentation for instructions on how to use it
%
% 	Speed of Light & $c_{0}$ & \SI{2.99792458e8}{\meter\per\second} (exact)\\
% %Constant Name & $Symbol$ & $Constant Value$ with units\\
%
% \end{constants}

%----------------------------------------------------------------------------------------
%	SYMBOLS
%----------------------------------------------------------------------------------------

% \begin{symbols}{lll} % Include a list of Symbols (a three column table)
%
% $a$ & distance & \si{\meter} \\
% $P$ & power & \si{\watt} (\si{\joule\per\second}) \\
% %Symbol & Name & Unit \\
%
% \addlinespace % Gap to separate the Roman symbols from the Greek
%
% $\omega$ & angular frequency & \si{\radian} \\
%
% \end{symbols}

%----------------------------------------------------------------------------------------
%	DEDICATION
%----------------------------------------------------------------------------------------

% \dedicatory{For/Dedicated to/To my\ldots}

%----------------------------------------------------------------------------------------
%	THESIS CONTENT - CHAPTERS
%----------------------------------------------------------------------------------------

\mainmatter % Begin numeric (1,2,3...) page numbering

\pagestyle{thesis} % Return the page headers back to the "thesis" style

% Include the chapters of the thesis as separate files from the Chapters folder
% Uncomment the lines as you write the chapters

\include{Chapters/Intro}
\include{Chapters/Rapmap}
\include{Chapters/Alevin}
\include{Chapters/Conclusion}
%\include{Chapters/Chapter4} 
%\include{Chapters/Chapter5} 

%----------------------------------------------------------------------------------------
%	THESIS CONTENT - APPENDICES
%----------------------------------------------------------------------------------------

\appendix % Cue to tell LaTeX that the following "chapters" are Appendices

% Include the appendices of the thesis as separate files from the Appendices folder
% Uncomment the lines as you write the Appendices

\include{Appendices/AppendixA}
\include{Appendices/AppendixB}
%\include{Appendices/AppendixC}

%----------------------------------------------------------------------------------------
%	BIBLIOGRAPHY
%----------------------------------------------------------------------------------------

\printbibliography[heading=bibintoc]

%----------------------------------------------------------------------------------------

\end{document}  
